android/app/src/main/java/org/mozilla/magnet/MagnetScannerReact.java

```diff
@@ -40,25 +56,98 @@ public String getName() {
      * Start the ScannerService scanning.
      */
     @ReactMethod
-    public void start() {
-        mMagnetScanner.start(this);
+    public void start(final Promise promise) {
+        Log.d(TAG, "start");
+
+        if (mStarted || mStarting) {
+            promise.reject("busy", "already starting");
+            return;
+        }
+
+        mStarting = true;
+        Log.d(TAG, "starting");
+
+        checkPermissions(new PermissionCheckerCallback() {
+            @Override
+            public void onPermissionChecksComplete() {
+                Log.d(TAG, "permission checks done");
+                mLastPermissionsCheck = System.currentTimeMillis();
+
+                // scanner can be stopped
+                // during permissions check
+                if (!mStarting) {
+                    promise.reject("stopped", "stopped before start finished");
+                    return;
+                }
+
+                mMagnetScanner.start(MagnetScannerReact.this);
+                mStarting = false;
+                mStarted = true;
+                promise.resolve(true);
+            }
+
+            @Override
+            public void onPermissionChecksError(Error error) {
+                Log.e(TAG, error.toString());
+                mStarting = false;
+                promise.reject("permission-error", error.getMessage());
+            }
+        });
     }
 
     /**
      * Stop the ScannerService scanning.
      */
     @ReactMethod
-    public void stop() {
+    public void stop(Promise promise) {
+        Log.d(TAG, "stop");
         mMagnetScanner.stop();
+        mStarting = false;
+        mStarted = false;
+        promise.resolve(true);
+    }
+
+    private void checkPermissions(PermissionCheckerCallback callback) {
+        if (!needsPermissionsCheck()) {
+            callback.onPermissionChecksComplete();
+            return;
+        }
+
+        Activity activity = mContext.getCurrentActivity();
+        if (activity == null) {
+            Log.e(TAG, "can't check when app in background");
+            callback.onPermissionChecksError(new Error("activity null"));
+            return;
+        }
+
+        mPermissionChecker.check(activity, callback);
+    }
+
+    private boolean needsPermissionsCheck() {
```

Comment:
I though the properties starting and started were going to be used to control that we don't need to check permissions again, but seems we are doing it based on time, don't you think this can cause races?